pip/pip-310.md

```diff
@@ -0,0 +1,121 @@
+* Produce path:
+  * ResourceGroupPublishLimiter - resource group's produce qps and MBps checks
+  * PrecisePublishLimiter - the topic quota rate limiter in case when `preciseTopicPublishRateLimiterEnable` broker config is `true`
+  * PublishRateLimiterImpl - the topic quota rate limiter in case when `preciseTopicPublishRateLimiterEnable` broker
+  config is `false`
+  * BrokerService - manage total publish rate of the broker
+* Consume path:
+  * DispatchRateLimiter - dispatch rate limiter
+  * BrokerService - manage total dispatch rate of the broker 
+* Misc:
+  * BrokerService - concurrent unloader rate limiter
+  * (Function|Sink|Source)StatsManager
+
+RateLimiter.java has two ways of implementing the rate limit check - precise and poller based. The poller based accumulates allowed permit and check for breach at a regular, configurable, frequency within each second. The precise one checks for breach before handing out permits each time.
+
+The DispatchRateLimiter hard codes RateLimiter to use precise rate limiter, but in the produce path , users can configure which one to choose for their brokers.
+
+# Motivation
+
+<!--
+Describe the problem this proposal is trying to solve.
+
+* Explain what is the problem you're trying to solve - current situation.
+* This section is the "Why" of your proposal.
+-->
+
+We have used both the poller based and precise publish rate limiters and there are challenges with both the approaches:
+
+* Poller based doesn't really do any throttling, all pending requests due to limit breach in the first second pass through in the first interval of the subsequent second.
+  * Effectively, there is no rate limiting.
+  * This also leads to noisy neighbour issues where topics hosted on the same broker/bookie face latency spikes.
+* Precise rate limiter fixes the above two issues, but introduces another challenge - the rate limiting is too strict.
+  * This leads potential for data loss in case there are sudden spikes in produce and client side produce queue breaches.
+  * The produce latencies increase exponentially in case produce breaches the set throughput even for small windows.
+
+Thus, there is a need for a rate limiter which is in between the above two implementations. As such, the handling of produce spikes
+and bursts may way from use-case to use-case, so it's not ideal to implement and add another rate limiter logic inside of Pulsar codebase
+as it will never be able to meet all possible requirements.
+
+# Goals
+
+I propose a configurable rate limiter implementation support, just like how AuthorizationProvider and
+AuthenticationProvider allow for.
+
+## In Scope
+
+For this PIP, the scope is the produce path only. It covers both the topic and the resource group level rate limiter.
+
+## Out of Scope
+
+Dispatch rate limiter is out of scope of this PIP. Generally speaking, the dispatch rate can smoothen out over time even if there are high produce spikes.
+
+# High Level Design
```

Comment:
I would start by describing what a Rate Limiter supposed to implement and in the Public API section I must provide the interface

Reply:
I thought, since its using an existing interface, i'd rather show the minimal delta, than the whole class which exists in the codebase already?

Reply:
Of course, but can you be more specific - which interface already exists for rate limiting? Just understand that if I'm asking those questions it means I couldn't answer them based on reading the doc. If a prequisite is to read certain parts of the code prior reading this PIP, it should be mentioned, but I think we can avoid that easily.